File: 0012_Integer_to_Roman/Integer_to_Roman.cpp
// Check py file for explaining comments


class Solution {
public:
    string intToRoman(int num) {
        string roman = ""; // Answer string
        int pos = 0;       // Position counter: 0=ones, 1=tens, 2=hundreds, 3=thousands

        unordered_map<int, string> dc = {
            {1, "I"}, {5, "V"}, {10, "X"}, {50, "L"},
            {100, "C"}, {500, "D"}, {1000, "M"}
        };

        while (num > 0) {
            int numLast = num % 10; // Last digit
            num /= 10;              // Remove last digit

            if (numLast == 0) {
                pos++;
                continue;
            }

            if (numLast < 4) {
                // 1, 2, 3
                string temp = "";
                for (int i = 0; i < numLast; i++) {
                    temp += dc[(int)pow(10, pos)];
                }
                roman = temp + roman;
            }
            else if (numLast == 4 || numLast == 9) {
                // 4, 9
                int numLastextra = numLast + 1;
                string temp = dc[(int)pow(10, pos)] +
                              dc[numLastextra * (int)pow(10, pos)];
                roman = temp + roman;
            }
            else { // 5, 6, 7, 8
                if (numLast == 5) {
                    string temp = dc[5 * (int)pow(10, pos)];
                    roman = temp + roman;
                } else {
                    int numLastextra = numLast - 5;
                    string temp = dc[5 * (int)pow(10, pos)];
                    for (int i = 0; i < numLastextra; i++) {
                        temp += dc[(int)pow(10, pos)];
                    }
                    roman = temp + roman;
                }
            }

            pos++;
        }

        return roman;
    }
};

File: 0012_Integer_to_Roman/Integer_to_Roman.py
class Solution:
    def intToRoman(self, num: int) -> str:
        roman = "" # Answer string
        pos = 0 # Position counter, 0 = ekok, 1 = dosok, 2 = sotok, 3 = hajar.
        dc = {1 : "I", 5 : "V", 10 : "X", 50 : "L", 100 : "C", 500 : "D", 1000 : "M"} # Milestone values
        while (num > 0):
            numLast = num%10 # Last digit katbe
            num = num // 10 # Ekta digit last theke komabo
            
            if numLast == 0:
                # Ei Line ta skip hobe
                pos += 1
                continue
            
            elif numLast < 4:
                # 1, 2, 3
                # 1 er jeta hobe setai baki duto er sathe gun hobe
                temp = dc[10**pos] * numLast
                roman = temp + roman
                
            elif numLast == 4 or numLast == 9:
                # 4, 9
                numLastextra = numLast + 1 # Milestone Value ta
                temp = dc[10**pos] + dc[numLastextra*(10**pos)] # Ek ghor kom er ta age bosbe then milestone value ta
                roman = temp + roman
            elif numLast < 9:
                # 5, 6, 7, 8
                if numLast == 5:
                    temp = dc[5*(10**pos)] # milestone value hole
                    roman = temp + roman
                else:
                    # 6, 7, 8
                    numLastextra = numLast - 5 # Milestone value bad dile je position er 1, 2, 3
                    temp =  dc[5*(10**pos)] + dc[10**pos] * numLastextra
                    roman = temp + roman

            pos += 1

        return roman

# Alternative Solution
# class Solution:
#     def intToRoman(self, num: int) -> str:
#         vals = [
#             (1000,"M"), (900,"CM"), (500,"D"), (400,"CD"),
#             (100,"C"), (90,"XC"), (50,"L"), (40,"XL"),
#             (10,"X"), (9,"IX"), (5,"V"), (4,"IV"), (1,"I")
#         ]
        
#         res = ""
#         for v, s in vals:
#             while num >= v:
#                 res += s
#                 num -= v
#         return res
